source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,809 @@
+              }();
+
+              const Point<dim> current_photon_location =
+                current_photon->get_location();
+
+              // Loop over the neighboring cells of the main cell. This includes
+              // the main cell itself.
+              auto &cell_neighbor_list = *cell_neighbor_list_iterator;
+              for (auto current_neighboring_cell = cell_neighbor_list.begin();
+                   current_neighboring_cell != cell_neighbor_list.end();
+                   ++current_neighboring_cell)
+                {
+                  // Particle iterators in the current neighboring cell
+                  typename Particles::ParticleHandler<
+                    dim>::particle_iterator_range particle_in_neighboring_cell =
+                    particle_handler.particles_in_cell(
+                      *current_neighboring_cell);
+
+                  // This vector will be reused for every line-sphere
+                  // intersection search. There is maximum 2 intersection points
+                  // between a line and a sphere.
+                  std::vector<Point<dim>> current_intersection_points;
+                  current_intersection_points.reserve(2);
+
+                  // Loop of the particles in the current neighboring cells
+                  for (auto current_particle =
+                         particle_in_neighboring_cell.begin();
+                       current_particle != particle_in_neighboring_cell.end();
+                       ++current_particle)
+                    {
+                      // Check if the current photon intersects the current
+                      // particle
+                      const Point<dim> particle_position =
+                        current_particle->get_location();
+                      const double particle_diameter =
+                        current_particle->get_properties()
+                          [DEMProperties::PropertiesIndex::dp];
+
+                      current_intersection_points =
+                        LetheGridTools::find_line_sphere_intersection(
+                          photon_insertion_point,
+                          photon_displacement_vector,
+                          particle_position,
+                          particle_diameter);
+
+                      // If the size of the vector is 0, there's no intersection
+                      // point, thus we go to the next particle.
+                      if (current_intersection_points.size() == 0)
+                        continue;
+
+                      // Otherwise, there is at least one intersection point.
+                      // Initializing the needed variables.
+                      double     new_distance;
+                      Point<dim> new_closest_point;
+
+                      // If the size of the vector is 1, we have one
+                      // intersection point.
+                      if (current_intersection_points.size() == 1)
```

Comment:
This if and else statement could be replaced by just a loop over the intersection points. It would make the code a lot easier to read and more streamlined.
Just initialized new_distance to be a large value

Just loop over the intersection points and take the min between new_distance and the distance of the current intersection point. This way you could have N intersection points and it would still work just fine